File: frontend/src/services/evidence.ts
/** Runtime validation for canonical evidence fields; prose is never a verified fact. */
import { z } from "zod";
import { apiClient } from "./api";
const value = z.union([z.number().finite(), z.string(), z.null()]);
export const evidenceSummarySchema = z.object({
  snapshot_ids: z.array(z.string()),
  dossier_ids: z.array(z.string()),
  errors: z.array(z.string()),
  prose_verification: z.literal("unverified"),
  actionable: z.literal(false),
});
export type EvidenceSummary = z.infer<typeof evidenceSummarySchema>;
const claim = z.object({
  kind: z.enum(["fact", "derived", "hypothesis", "judgment"]),
  symbol: z.string(),
  metric: z.string(),
  value,
  unit: z.string(),
  period: z.string(),
  period_end: z.string().nullable(),
  evidence_ids: z.array(z.string()),
  method: z.string().nullable(),
  text: z.string(),
});
const result = z.object({
  claim_id: z.string(),
  claim,
  status: z.enum(["matches_snapshot", "unverified", "rejected"]),
  reasons: z.array(z.string()),
  evidence_ids: z.array(z.string()),
  computed_value: z.number().finite().nullable(),
});
const dossierSchema = z.object({
  dossier: z.object({
    dossier_id: z.string(),
    snapshot_id: z.string(),
    manifest_hash: z.string(),
    run_id: z.string(),
    symbol: z.string(),
    claims: z.array(result),
    errors: z.array(z.string()),
    prose_verification: z.literal("unverified"),
    actionable: z.literal(false),
  }),
  run_status: z.string().nullable(),
});
export type Dossier = z.infer<typeof dossierSchema>;
const recordSchema = z.object({
  evidence_id: z.string(),
  snapshot_id: z.string(),
  symbol: z.string(),
  instrument_id: z.string(),
  metric: z.string(),
  value,
  unit: z.string(),
  period: z.string(),
  period_start: z.string().nullable(),
  period_end: z.string().nullable(),
  observed_at: z.string().nullable(),
  published_at: z.string().nullable(),
  fetched_at: z.string(),
  provider: z.string(),
  source_uri: z.string().nullable(),
  document_id: z.string().nullable(),
  point_in_time_status: z.enum(["verified", "retrieval_only", "unknown"]),
  quality: z.enum([
    "available",
    "stale",
    "missing",
    "conflicting",
    "unsupported",
  ]),
  payload_hash: z.string(),
  adapter_version: z.string(),
});
export type EvidenceRecord = z.infer<typeof recordSchema>;
const snapshotSchema = z.object({
  snapshot_id: z.string(),
  state: z.literal("sealed"),
  manifest_hash: z.string(),
  coverage: z.record(z.string()),
  conflicts: z.record(z.array(z.string())),
  records_count: z.number().int(),
});
export async function getRunEvidence(
  run: string,
  signal?: AbortSignal,
): Promise<EvidenceSummary | null> {
  return evidenceSummarySchema
    .nullable()
    .parse(
      (
        await apiClient.get<unknown>(
          `/api/portfolio/evidence/runs/${encodeURIComponent(run)}`,
          { signal },
        )
      ).data,
    );
}
export async function getDossier(
  id: string,
  signal?: AbortSignal,
): Promise<Dossier> {
  return dossierSchema.parse(
    (
      await apiClient.get<unknown>(
        `/api/portfolio/evidence/dossiers/${encodeURIComponent(id)}`,
        { signal },
      )
    ).data,
  );
}
export async function getEvidence(
  id: string,
  snapshot: string,
  signal?: AbortSignal,
): Promise<EvidenceRecord> {
  return recordSchema.parse(
    (
      await apiClient.get<unknown>(
        `/api/portfolio/evidence/records/${encodeURIComponent(id)}`,
        { params: { snapshot_id: snapshot }, signal },
      )
    ).data,
  );
}
export async function getManifest(id: string, signal?: AbortSignal) {
  return snapshotSchema.parse(
    (
      await apiClient.get<unknown>(
        `/api/portfolio/evidence/snapshots/${encodeURIComponent(id)}`,
        { params: { limit: 1 }, signal },
      )
    ).data,
  );
}

File: backend/src/models/evidence.py
"""Versioned evidence identities, sealed manifests and structured-field validation results."""

from datetime import date, datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .research_strategy import StrategyVersion

Number = Annotated[float, Field(strict=True, allow_inf_nan=False)]
Family = Literal[
    "quote",
    "ohlcv",
    "overview",
    "cash_flow",
    "balance_sheet",
    "news",
    "filing",
    "income_statement",
]
Quality = Literal["available", "stale", "missing", "conflicting", "unsupported"]
Period = Literal["instant", "session", "quarter", "annual", "ttm", "event", "unknown"]


class EvidenceModel(BaseModel):
    model_config = ConfigDict(
        extra="forbid", allow_inf_nan=False, revalidate_instances="always"
    )


class EvidenceRecord(EvidenceModel):
    evidence_id: str = ""
    snapshot_id: str
    symbol: str = Field(pattern=r"^[A-Z0-9][A-Z0-9.-]{0,14}$")
    instrument_id: str
    exchange: str | None = None
    currency: str | None = None
    family: Family
    metric: str = Field(max_length=80)
    value: Number | str | None = None
    unit: str = Field(max_length=40)
    period: Period = "unknown"
    period_start: date | None = None
    period_end: date | None = None
    observed_at: datetime | None = None
    session: str | None = None
    published_at: datetime | None = None
    fetched_at: datetime
    provider: str = Field(max_length=80)
    source_uri: str | None = None
    document_id: str | None = None
    point_in_time_status: Literal["verified", "retrieval_only", "unknown"] = "unknown"
    quality: Quality = "available"
    adjustment: str = "unknown"
    payload_hash: str = ""
    adapter_version: Literal["idq-004@1", "idq-005@1"] = "idq-004@1"
    revision_of: str | None = None
    legacy_aliases: list[str] = Field(default_factory=list, max_length=5)

    @model_validator(mode="after")
    def honest_availability(self) -> "EvidenceRecord":
        if self.quality == "available" and (
            self.value is None or self.unit == "unknown"
        ):
            raise ValueError("Available evidence requires a value and a known unit")
        if self.point_in_time_status == "verified" and (
            self.published_at is None or self.document_id is None
        ):
            raise ValueError("PIT proof requires publication and document identity")
        if (
            self.period_start
            and self.period_end
            and self.period_start > self.period_end
        ):
            raise ValueError("Reversed period")
        for moment in [self.fetched_at, self.observed_at, self.published_at]:
            if moment is not None and moment.tzinfo is None:
                raise ValueError("Evidence times must be timezone aware")
        return self


EXPECTED_FAMILIES: tuple[Family, ...] = (
    "quote",
    "ohlcv",
    "overview",
    "cash_flow",
    "balance_sheet",
    "news",
    "filing",
)


class EvidenceSnapshot(EvidenceModel):
    schema_version: Literal[1] = 1
    freshness_policy: Literal["diagnostic-close@1", "strategy-contract@1"] = (
        "diagnostic-close@1"
    )
    reconciliation_policy: Literal["like-slot-1e-6@1"] = "like-slot-1e-6@1"
    source_selection: Literal["retain-all-comparable-no-average@1"] = (
        "retain-all-comparable-no-average@1"
    )
    snapshot_id: str
    request_hash: str
    run_id: str
    symbol: str
    requested_as_of: datetime
    mode: Literal["research", "strict_historical"] = "research"
    state: Literal["collecting", "sealed", "failed", "cancelled"] = "collecting"
    generation: int = 1
    owner: str
    lease_until: datetime
    created_at: datetime
    sealed_at: datetime | None = None
    parent_id: str | None = None
    risk_snapshot_id: str | None = None
    policy_revision: int | None = None
    strategy_version: str | None = None
    strategy_contract: StrategyVersion | None = None
    strategy_peers: dict[str, str] = Field(default_factory=dict)
    records: list[EvidenceRecord] = Field(default_factory=list, max_length=1000)
    expected: list[Family] = Field(default_factory=lambda: list(EXPECTED_FAMILIES))
    coverage: dict[str, Quality] = Field(default_factory=dict)
    conflicts: dict[str, list[str]] = Field(default_factory=dict)
    manifest_hash: str = ""
    error_code: str | None = None


class ResearchClaim(EvidenceModel):
    kind: Literal["fact", "derived", "hypothesis", "judgment"]
    symbol: str
    metric: str
    value: Number | str | None = None
    unit: str = "unknown"
    period: Period = "unknown"
    period_end: date | None = None
    evidence_ids: list[str] = Field(default_factory=list, max_length=10)
    method: Literal["free_cash_flow@1", "net_margin@1"] | None = None
    text: str = Field(default="", max_length=1000)


class ClaimResult(EvidenceModel):
    claim_id: str
    claim: ResearchClaim
    status: Literal["matches_snapshot", "unverified", "rejected"]
    reasons: list[str]
    evidence_ids: list[str]
    computed_value: Number | None = None


class ResearchDossier(EvidenceModel):
    schema_version: Literal[1] = 1
    dossier_id: str
    snapshot_id: str
    manifest_hash: str
    run_id: str
    symbol: str
    report_hash: str
    claims: list[ClaimResult] = Field(max_length=64)
    errors: list[str]
    prose_verification: Literal["unverified"] = "unverified"
    actionable: Literal[False] = False


class EvidenceSummary(EvidenceModel):
    snapshot_ids: list[str]
    dossier_ids: list[str]
    errors: list[str]
    prose_verification: Literal["unverified"] = "unverified"
    actionable: Literal[False] = False
